R/calc_truth.R

```diff
@@ -0,0 +1,342 @@
+#' data set and calculates additional true information such as converting
+#' monthly age-composition information into annual data. No sampling is
+#' performed just mathematical operations.
+#' 
+#' @details The following values are calculated within this function:
+#'   * biomass
+#'   * catch
+#'   * numbers-at-age by dividing biomass-at-age by weight-at-age information
+#'   * weight
+#' 
+#' @param data A tibble containing ecosystem model (e.g., Ecopath with Ecosim) 
+#' outputs from [ecosystemdata::load_model()].
+#' @param species_name A string specifying the species name to subset.
+#' 
+#' @return
+#' A nested tibble where each row holds an output type (e.g., biomass, catch, 
+#' numbers, or weight) for a given time step category (e.g., monthly or yearly).
+#' Thus, data within a row in returned output can be expanded to reveal the
+#' full data, e.g., all months for an output type.
+#' 
+#' @examples
+#' data(ewe_nwatlantic_base, package = "ecosystemdata")
+#' data <- ewe_nwatlantic_base
+#' truth <- calc_truth(
+#'   data,
+#'   species_name = "menhaden"
+#' )
+#' 
+#' # To unnest tibble and view the data
+#' truth_long <- tidyr::unnest(truth)
+#'   
+#' # To calculate truth for a vector of species using {purrr}
+#' species_names <- c("menhaden", "striped bass", "bluefish")
+#' truth <- purrr::map_dfr(
+#'   species_names,
+#'   function(x) calc_truth(data, species_name = x)
+#' )
+#' @export
+#' @keywords calc_truth
+calc_truth <- function(data, species_name){
+
+  # Check if the species_name is in the data
+  unique_species <- unique(data[["species"]])
+  if (!species_name %in% unique_species) {
+    cli::cli_abort(c(
+      "x" = "Species {.val {species_name}} not found in the data.",
+      "i" = "Available species are: {.val {unique_species}}",
+      "i" = "Please check the species name and try again."
+    ))
+  }
+
+  # Create a tibble with the labels (e.g., biomass) and types (e.g., index)
+  # Define available labels from data
+  labels <- c("biomass", "catch", "weight")
+  unique_data_type <- unique(data[["type"]])
+  missing_data <- setdiff(labels, unique_data_type)
+  if (length(missing_data) > 0) {
+    cli::cli_abort(c(
+      "x" = "{.val {missing_data}}-at-age data is missing for {species_name}.",
+      "i" = "Available types in the data are: {.val {unique_data_type}}",
+      "i" = "Please ensure that the data contains {missing_data} information."
+    ))
+  }
+
+  # Create a tibble that maps labels to their types (index or age composition)
+  types <- dplyr::tibble(
+    label = rep(labels, 2),
+    type = c(
+      # weight only has agecomp type, NA for weight index
+      ifelse(labels %in% c("biomass", "catch"), "index", NA_character_),
+      rep("agecomp", length(labels))
+    )
+  ) |>
+    dplyr::filter(!is.na(type))
+
+  # Extract monthly data by labels
+  truth_monthly <- purrr::map(
+    labels,
+    function(x) calc_truth_monthly(data, species_name, x)
```

Comment:
This is assuming that all EWE models are reported in monthly time steps. Is that true?

Reply:
@rklasky, does the Ecosim output in CSV format always include both monthly and yearly time step results, regardless of the time unit set in Ecopath -> Input -> Model Parameters -> Time units (e.g., year, day, or another unit)? What about Ecospace? Sorry for not testing this myself. My version of EwE is outdated and currently in read-only mode.